### PR TITLE
Add fcos, fsaf, ssd head unittest case and fix fcos deploy bug

### PR DESCRIPTION
**Pull request by QingChuanWS**
This pr adds unit tests for FCOS, FSAF, and SSD Head, and fixes the problems found by FCOS when it passes this unit test.

### CODE REVIEW

File: tests/test_onnx/test_head.py [hunk 1]
```diff
@@ -224,3 +215,289 @@ def test_yolov3_head_get_bboxes():
     for i in range(len(onnx_outputs)):
         np.testing.assert_allclose(
             torch_outputs[i], onnx_outputs[i], rtol=1e-03, atol=1e-05)
+
+
+def fcos_config():
+    """FCOS Head Config."""
+    test_cfg = mmcv.Config(
+        dict(
+            deploy_nms_pre=1000,
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:224]: try to test whether `deploy_nms_pre` config is valid in the testing.

Comment (QingChuanWS) [reply tests/test_onnx/test_head.py:224]: All head unittest case make the `deploy_nms_pre` valid.

File: tests/test_onnx/test_head.py [hunk 2]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+            deploy_nms_pre=0,
+            min_bbox_size=0,
+            score_thr=0.05,
+            nms=dict(type='nms', iou_threshold=0.5),
+            max_per_img=100))
+
+    model = FCOSHead(num_classes=4, in_channels=1, test_cfg=test_cfg)
+
+    model.requires_grad_(False)
+    model.eval()
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:234]: change to cpu : `wrap_model.cpu().eval()`

File: tests/test_onnx/test_head.py [hunk 3]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+        dict(
+            deploy_nms_pre=0,
+            min_bbox_size=0,
+            score_thr=0.05,
+            nms=dict(type='nms', iou_threshold=0.5),
+            max_per_img=100))
+
+    model = FCOSHead(num_classes=4, in_channels=1, test_cfg=test_cfg)
+
+    model.requires_grad_(False)
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:233]: Why set to `False` for model?

File: tests/test_onnx/test_head.py [hunk 4]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+
+
+def test_fcos_head_forward_single():
+    """Test fcos forward single in torch and ort env."""
+    fcos_model = fcos_config()
+
+    feat = torch.rand(1, fcos_model.in_channels, 32, 32)
+    fcos_model.forward_single = partial(
+        fcos_model.forward_single,
+        scale=Scale(1.0).requires_grad_(False),
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:245]: Why set to False for requires_grad?

File: tests/test_onnx/test_head.py [hunk 5]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+    s = 128
+    feats = [
+        torch.rand(1, 1, s // feat_size, s // feat_size)
+        for feat_size in [4, 8, 16, 32, 64]
+    ]
+    ort_validate(fcos_model.forward, feats)
+
+
+def test_fcos_head_get_bboxes():
+    """Test fcos head get_bboxes() in ort and."""
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:264]: `"""Test fcos head get_bboxes() in ort"""`

File: tests/test_onnx/test_head.py [hunk 6]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+    ]
+
+    fcos_model.get_bboxes = partial(
+        fcos_model.get_bboxes, img_metas=img_metas, with_nms=False)
+    wrap_model = WrapFunction(fcos_model.get_bboxes)
+    wrap_model.cpu().eval()
+    with torch.no_grad():
+        torch.onnx.export(
+            wrap_model, (cls_scores, bboxes, centerness),
+            'tmp.onnx',
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:295]: This file `'tmp.onnx'` should be removed.

File: tests/test_onnx/test_head.py [hunk 7]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+    centerness = [
+        torch.rand(1, 1, s // feat_size, s // feat_size)
+        for feat_size in [4, 8, 16, 32, 64]
+    ]
+
+    fcos_model.get_bboxes = partial(
+        fcos_model.get_bboxes, img_metas=img_metas, with_nms=False)
+    wrap_model = WrapFunction(fcos_model.get_bboxes)
+    wrap_model.cpu().eval()
+    with torch.no_grad():
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:292]: Why not use `ort_validate` here?

File: tests/test_onnx/test_head.py [hunk 8]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+    # the bboxes's size: (1, 4, 64, 64), (1, 4, 32, 32),
+    # (1, 4, 16, 16), (1, 4, 8, 8), (1, 4, 4, 4).
+    fsaf_head_data = 'fsaf_head_get_bboxes.pkl'
+    feats = mmcv.load(osp.join(data_path, fsaf_head_data))
+    cls_score = feats[:5]
+    bboxes = feats[5:]
+
+    fsaf_model.get_bboxes = partial(
+        fsaf_model.get_bboxes, img_metas=img_metas, with_nms=False)
+    wrap_model = WrapFunction(fsaf_model.get_bboxes)
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:386]: Why not use `ort_validate` here?

File: tests/test_onnx/test_head.py [hunk 9]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+    wrap_model.cpu().eval()
+    with torch.no_grad():
+        torch.onnx.export(
+            wrap_model, (cls_score, bboxes),
+            'tmp.onnx',
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:391]: At the end of function `ort_validate`, if file `onnx_io` exists, it should be removed.

File: tests/test_onnx/test_head.py [hunk 10]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+            max_per_img=200))
+
+    model = SSDHead(
+        num_classes=4,
+        in_channels=(4, 8, 4, 2, 2, 2),
+        test_cfg=test_cfg,
+        **cfg)
+
+    model.requires_grad_(False)
+    model.eval()
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:442]: only have to move to cpu here once.

File: tests/test_onnx/test_head.py [hunk 11]
```diff
@@ -224,3 +216,292 @@ def test_yolov3_head_get_bboxes():
...
+    ssd_head_data = 'ssd_head_get_bboxes.pkl'
+    feats = mmcv.load(osp.join(data_path, ssd_head_data))
+    cls_score = feats[:6]
+    bboxes = feats[6:]
+
+    ssd_model.get_bboxes = partial(
+        ssd_model.get_bboxes, img_metas=img_metas, with_nms=False)
+    wrap_model = WrapFunction(ssd_model.get_bboxes)
+    wrap_model.cpu().eval()
+    with torch.no_grad():
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:486]: Why not use ort_validate here?

File: tests/test_onnx/utils.py [hunk 12]
```diff
@@ -27,15 +27,16 @@ def forward(self, *args, **kwargs):
         return self.wrapped_function(*args, **kwargs)
 
 
-def ort_validate(model, feats, onnx_io='tmp.onnx'):
+def ort_validate(model_func, feats, onnx_io='tmp.onnx'):
```
Comment (RunningLeon) [inline tests/test_onnx/utils.py:30]: here input could be a normal function or an instance of torch.nn.Module.

File: tests/test_onnx/utils.py [hunk 13]
```diff
@@ -27,15 +27,16 @@ def forward(self, *args, **kwargs):
...
     """Validate the output of the onnxruntime backend is the same as the output
     generated by torch.
 
     Args:
-        model (nn.Module): the model to be verified
+        model_func : the function of model to be verified
         feats (list(torch.Tensor) | torch.Tensor): the input of model
         onnx_io (str): the name of onnx output file
     """
+    model = WrapFunction(model_func)
```
Comment (RunningLeon) [inline tests/test_onnx/utils.py:39]: Use isinstance to check if input is a function or instance of torch.nn.Module

Comment (RunningLeon) [reply tests/test_onnx/utils.py:39]: And remove the onnx_io file at the end of this function if the file exists.

File: tests/test_onnx/test_head.py (hunk 3 above)
Comment (QingChuanWS) [reply tests/test_onnx/test_head.py:184]: set `False` for `requires_grad` is to freeze the model parameters

Comment (RunningLeon) [reply tests/test_onnx/test_head.py:184]: But you already use `with torch.no_grad()` in `ort_validate`.  Please verify if it is necessary to explicitly set `requires_grad` to `False`.

File: tests/test_onnx/test_head.py [hunk 14]
```diff
@@ -199,28 +164,223 @@ def test_yolov3_head_get_bboxes():
...
+    ]
+    ort_validate(fsaf_model.forward, feats)
+
+
+def test_fsaf_head_get_bboxes():
+    """Test RetinaNet Head get_bboxes in torch and onnxruntime env."""
+    fsaf_model = fsaf_config()
+    s = 256
+    img_metas = [{
+        'img_shape_for_onnx': (s, s, 3),
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:292]: `img_shape_for_onnx` should be a tensor of [h,w]

File: tests/test_onnx/test_head.py [hunk 15]
```diff
@@ -199,28 +164,223 @@ def test_yolov3_head_get_bboxes():
...
+    ]
+    ort_validate(ssd_model.forward, feats)
+
+
+def test_ssd_head_get_bboxes():
+    """Test SSD Head get_bboxes in torch and onnxruntime env."""
+    ssd_model = ssd_config()
+    s = 300
+    img_metas = [{
+        'img_shape_for_onnx': (s, s, 3),
```
Comment (RunningLeon) [inline tests/test_onnx/test_head.py:366]: `img_shape_for_onnx` should be a tensor of [h,w]

File: tests/test_onnx/utils.py [hunk 16]
```diff
@@ -32,14 +33,18 @@ def ort_validate(model, feats, onnx_io='tmp.onnx'):
...
-        feats (list(torch.Tensor) | torch.Tensor): the input of model
-        onnx_io (str): the name of onnx output file
+        model (nn.Module | function) : the function of model or model
+        to be verified.
+        feats (list(torch.Tensor) | torch.Tensor): the input of model.
+        onnx_io (str): the name of onnx output file.
     """
-    model.cpu().eval()
+    # if mdoel is a function, wrapped the function.
+    if (hasattr(model, '__call__')):
```
Comment (RunningLeon) [inline tests/test_onnx/utils.py:42]: Remove outside `()`

Comment (RunningLeon) [reply tests/test_onnx/utils.py:42]:
```
else:
       wrap_model = model
```

File: tests/test_onnx/utils.py [hunk 17]
```diff
@@ -32,14 +33,18 @@ def ort_validate(model, feats, onnx_io='tmp.onnx'):
...
+        wrap_model = WrapFunction(model)
```
Comment (RunningLeon) [inline tests/test_onnx/utils.py:43]:
```
else:
       wrap_model = model
```

File: tests/test_onnx/utils.py (hunk 16 above)
Comment (RunningLeon) [reply tests/test_onnx/utils.py:42]: Verify if `hasattr(model, '__call__')` always return `True ` for normal function, instance of an `torch.nn.Module` or the `forward` method of an instance of an `torch.nn.Module`.

File: tests/test_onnx/utils.py [hunk 18]
```diff
@@ -48,9 +53,21 @@ def ort_validate(model, feats, onnx_io='tmp.onnx'):
...
             opset_version=11)
 
-    onnx_outputs = verify_model(feats)
+    if isinstance(feats, tuple):
+        ort_feats = []
+        for feat in feats:
+            ort_feats += feat
+    else:
+        ort_feats = feats
+    onnx_outputs = verify_model(ort_feats)
```
Comment (RunningLeon) [inline tests/test_onnx/utils.py:62]: `verify_model` function basically only gets the onnx outputs, which means the name `verify_model` is misleading. Please change to a more self-explaining name for this function.

File: tests/test_onnx/utils.py [hunk 19]
```diff
@@ -32,14 +33,22 @@ def ort_validate(model, feats, onnx_io='tmp.onnx'):
...
-        feats (list(torch.Tensor) | torch.Tensor): the input of model
-        onnx_io (str): the name of onnx output file
+        model (nn.Module | function) : the function of model or model
+            to be verified.
+        feats (tuple(list(torch.Tensor)) | list(torch.Tensor) | torch.Tensor):
+            the input of model.
+        onnx_io (str): the name of onnx output file.
     """
-    model.cpu().eval()
+    # if mdoel is not a nn.Module, then it is a function and wrapped
```
Comment (RunningLeon) [inline tests/test_onnx/utils.py:42]: if model is not an instance of nn.Module, then it is a normal function and it should be wrapped.

**Comment by RunningLeon** [comment]
@ZwwWayne LGTM, see if you have any comments.

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**